### PR TITLE
Use `Circuit` for Bittide Nodes

### PR DESCRIPTION
**Pull request by lmbollen**
First effort to use `Circuit` in Bittide nodes

### CODE REVIEW

File: bittide/src/Bittide/Switch.hs [hunk 1]
```diff
@@ -45,15 +33,18 @@ mkSwitch ::
...
-  Vec links (Signal dom (DataLink frameWidth)) ->
-  ( Vec links (Signal dom (DataLink frameWidth))
-  , Vec (1 + (links * 2)) (Signal dom (WishboneS2M (Bytes nBytes)))
-  )
-mkSwitch SwitchConfig{..} = switch preamble calendarConfig
+  CalendarConfig nBytes addrW (CalendarEntry links) ->
+  Circuit
+    ( CSignal dom (Vec links (DataLink frameWidth))
+    , Wishbone dom 'Standard addrW (Bytes nBytes)
+    )
```
Comment (cuddlefishie) [inline bittide/src/Bittide/Switch.hs:40]:
```suggestion
  Circuit
    ( CSignal dom (Vec links (DataLink frameWidth))
    , Wishbone dom 'Standard addrW (Bytes nBytes) -- calendar interface
    )
```

File: bittide/src/Bittide/Node.hs [hunk 2]
```diff
@@ -196,15 +202,34 @@ managementUnit ::
...
+--   Circuit
+--     (CSignal dom (DataLink 64))
+--     (CSignal dom (DataLink 64), Vec (32 - CLog 2 (nodeBusses + 8)) (Wishbone dom 'Standard 4 (Bytes 4)))
+-- managementUnitC (GppeConfig peConfig scatterConfig gatherConfig) = circuit $ \(linkIn, nmuWbs) -> do
+--   [wbScatCal, wbScat, wbGathCal, wbGu] -< nmuWbs
+--   jtag <- idleSource -< ()
+--   peWbs <- processingElement peConfig -< jtag
+--   linkOut <- gatherUnitWb gatherConfig -< (wbGathCal, wbGu)
+--   scatterUnitWb scatterConfig -< (linkIn, wbScatCal, wbScat)
+--   idC -< linkOut
```
Comment (cuddlefishie) [inline bittide/src/Bittide/Node.hs:235]: Is this not compiling yet? I'll give it a try locally, just commenting here so I don't forget

**Review by cuddlefishie — COMMENTED** [review]
Overall looks good, could use a rebase since the Link stuff was removed and some comments should be updated.

File: bittide/src/Bittide/Node.hs [hunk 3]
```diff
@@ -4,25 +4,30 @@
...
 import Bittide.ScatterGather
 import Bittide.SharedTypes
 import Bittide.Switch
 
-import Control.Arrow (first, (&&&), (>>>))
+import Control.Arrow ((&&&))
+
+instance IdleCircuit (Jtag dom) where
+  idleFwd _ = pure $ JtagIn low low low
+  idleBwd _ = pure $ JtagOut low low
```
[the remark this answers is not in the document]
Comment (lmbollen) [reply bittide/src/Bittide/Node.hs:30]: https://github.com/clash-lang/clash-vexriscv/pull/36 :)

File: bittide/src/Bittide/Node.hs [hunk 4]
```diff
@@ -191,3 +210,112 @@ managementUnit (ManagementConfig scatterConfig gatherConfig peConfig) linkIn nod
...
+  ([wbScatCal, wbScat, wbGathCal, wbGu], nmuWbs) <- splitAtC d4 -< peWbs
+  linkOut <- gatherUnitWbC gatherConfig -< (wbGu, wbGathCal)
+  scatterUnitWbC scatterConfig -< (linkIn, wbScat, wbScatCal)
+  idC -< (linkOut, nmuWbs)
+
+-- These functions should be added to `clash-protocols`, there is a PR for this:
+-- https://github.com/clash-lang/clash-protocols/pull/116
+-- And a bittide-hardware issue:
+-- https://github.com/bittide/bittide-hardware/issues/645
+-- Append two separate vectors of the same circuits into one vector of circuits
```
Comment (hydrolarus) [inline bittide/src/Bittide/Node.hs:242]: The PR on `clash-protocols` is still in draft, should that be merged in before this PR or do we merge this first, given this PR is open?

**Review by hydrolarus — APPROVED** [review]
Looking good! Just a question about PR order stuff

File: bittide/src/Bittide/Node.hs (hunk 4 above)
Comment (lmbollen) [reply bittide/src/Bittide/Node.hs:242]: Let's not block this PR on that, there's a tracking issue and I _intend_ to get back to that in the near future :)

**Pull request merged**